lib/simulateSpendCondTxLocally.js

```diff
@@ -0,0 +1,289 @@
+    },
+  });
+
+  return new Promise((resolve, reject) => {
+    // run the tx \o/
+    vm.runTx({
+      tx
+    }, (err, results) => {
+      if (err) {
+        return reject(err);
+      }
+      if (results.vm.exceptionError) {
+        return reject(results.vm.exceptionError);
+      }
+      return resolve(results);
+    });
+  });
+}
+
+function aggregateLogOutsByColorFromTo(logOuts) {
+  const outsGroupedByColor = {}
+  logOuts.forEach(item => {
+    if (outsGroupedByColor[item.color]) {
+      outsGroupedByColor[item.color].push({ from: item.from, to: item.to, amount: item.amount })
+    }
+    else {
+      outsGroupedByColor[item.color] = [{ from: item.from, to: item.to, amount: item.amount }]
+    }
+  })
+
+  const outsGroupedByColorFrom = {}
+  Object.keys(outsGroupedByColor).forEach(color => {
+    outsGroupedByColorFrom[color] = {}
+  })
+  Object.keys(outsGroupedByColor).forEach(color => {
+    outsGroupedByColor[color].forEach(out => {
+      if (outsGroupedByColorFrom[color][out.from]) {
+        // aggregate based on same color and from address
+        const ii = outsGroupedByColorFrom[color][out.from].findIndex(item => item.to === out.to);
+        if (ii !== -1) {
+          outsGroupedByColorFrom[color][out.from][ii] = { to: out.to, amount: add(out.amount, outsGroupedByColorFrom[color][out.from][ii].amount) }
+        } else{
+          outsGroupedByColorFrom[color][out.from].push({ to: out.to, amount: out.amount })
+        }
+      } else {
+        outsGroupedByColorFrom[color][out.from] = [{ to: out.to, amount: out.amount }]
+      }
+    }
+    )
+  })
+  return outsGroupedByColorFrom
+}
+
+function calcOutputs(unspent, inputs, aggregatedOuts) {
+  if (unspent.length === 0) {
+    throw new Error('Unspent is empty');
+  }
+  const outputs = []
+  Object.keys(aggregatedOuts).forEach(color => {
+
+    // sum total unspent amount of one color from inputs
+    const inInputs = u => {
+      return inputs.findIndex(input =>
+        u.outpoint.hash.equals(input.prevout.hash) &&
+        (u.outpoint.index === input.prevout.index)) > -1;
+    }
+
+    const isColor = u => color === String(u.output.color);
+    let sum = unspent.filter(isColor).filter(inInputs).reduce((a, u) => add(a, BigInt(u.output.value)), BigInt(0));
+
+    Object.keys(aggregatedOuts[color]).forEach(from => {
+      aggregatedOuts[color][from].forEach(out => {
+        if (lessThan(sum, BigInt(out.amount))) {
+          throw new Error('Not enough inputs');
+        }
+        sum = subtract(sum, BigInt(out.amount));
+        outputs.push(new Output(out.amount, out.to.toLowerCase(), color));
+      })
+      // generate refund output for input
+      if (greaterThan(sum, BigInt(0))) {
+        outputs.push(new Output(sum, from.toLowerCase(), color));
+      }
+    })
+  })
+  return outputs;
+}
+
+// user should provide erc20 addresses used in tx's spend condition contracts,  
+module.exports = async (tx, erc20sToBeUsed, state) => {
```

Comment:
The `state` coulb be replaced with the actual used `unspent[]` here

Reply:
yeah, we can simplify this augument